### PR TITLE
Mx 14324 sovereing refactor factories 2

### PR DESCRIPTION
**Pull request by raduchis**
## Reasoning behind the pull request
- implementation of factories for sovereign and normal
  
## Proposed changes
- added factories for creation of sovereign and normal components

## Testing procedure
- allIn; nothing should be broken

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

**Review by sasurobert — APPROVED** [review]

File: process/track/sovereignBlockTrackerFactory.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package track
+
+import (
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-go/process"
+)
+
+// BlockTrackerCreator is an interface for creating block trackers
+type BlockTrackerCreator interface {
```
Comment (SebastianMarian) [inline process/track/sovereignBlockTrackerFactory.go:9]: Could be this moved in process/interface.go or process/track/interface.go?

File: process/track/sovereignBlockTrackerFactory.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	CreateBlockTracker(argBaseTracker ArgShardTracker) (process.BlockTracker, error)
+	IsInterfaceNil() bool
+}
+
+type sovereignBlockTrackerFactory struct {
+	blockTrackerCreator BlockTrackerCreator
+}
+
+// NewSovereignBlockTrackerFactory creates a new shard block tracker factory for sovereign chain
+func NewSovereignBlockTrackerFactory(btc BlockTrackerCreator) (*sovereignBlockTrackerFactory, error) {
```
Comment (SebastianMarian) [inline process/track/sovereignBlockTrackerFactory.go:19]: PoS instead PoW? eth or egld instead btc :)

File: process/coordinator/sovereignTransactionCoordinatorFactory.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
+package coordinator
+
+import (
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-go/process"
+)
+
+type TransactionCoordinatorCreator interface {
```
[the remark this answers is not in the document]
Comment (raduchis) [reply process/coordinator/sovereignTransactionCoordinatorFactory.go:8]: moved to process/coordinator/interface.go

File: process/coordinator/shardTransactionCoordinatorFactory.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+type shardTransactionCoordinatorFactory struct {
+}
+
+// NewShardTransactionCoordinatorFactory creates a new sovereign transaction coordinator factory
+func NewShardTransactionCoordinatorFactory() (*shardTransactionCoordinatorFactory, error) {
+	return &shardTransactionCoordinatorFactory{}, nil
+}
+
+// CreateTransactionCoordinator creates a new sovereign transaction coordinator for the chain run type sovereign
```
[the remark this answers is not in the document]
Comment (raduchis) [reply process/coordinator/shardTransactionCoordinatorFactory.go:13]: renamed without sovereign

File: process/block/shardHeaderValidatorFactory.go [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+type shardHeaderValidatorFactory struct {
+}
+
+// NewShardHeaderValidatorFactory creates a new shard header validator factory
+func NewShardHeaderValidatorFactory() (*shardHeaderValidatorFactory, error) {
+	return &shardHeaderValidatorFactory{}, nil
+}
+
+// CreateHeaderValidator creates a new header validator for the chain run type shard
```
[the remark this answers is not in the document]
Comment (raduchis) [reply process/block/shardHeaderValidatorFactory.go:13]: yes, changed

File: factory/processing/interface.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+// HeaderValidatorCreator is an interface for creating header validators
+type HeaderValidatorCreator interface {
+	CreateHeaderValidator(args block.ArgsHeaderValidator) (process.HeaderConstructionValidator, error)
+	IsInterfaceNil() bool
+}
+
+// BlockTrackerCreator is an interface for creating block trackers
+type BlockTrackerCreator interface {
+	CreateBlockTracker(argBaseTracker track.ArgShardTracker) (process.BlockTracker, error)
```
[the remark this answers is not in the document]
Comment (raduchis) [reply factory/processing/interface.go:39]: renamed  to argShardTracker

File: process/sync/sovereignForkDetectorFactory.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+// CreateForkDetector creates a new fork detector
+func (s *sovereignForkDetectorFactory) CreateForkDetector(args ForkDetectorFactoryArgs) (process.ForkDetector, error) {
+	sfd, err := s.shardForkDetectorFactory.CreateForkDetector(args)
+	if err != nil {
+		return nil, err
+	}
+
+	sharFD, ok := sfd.(*shardForkDetector)
```
[the remark this answers is not in the document]
Comment (raduchis) [reply process/sync/sovereignForkDetectorFactory.go:35]: renamed to sfd and fd

File: process/track/sovereignBlockTrackerFactory.go (hunk 1 above)
Comment (raduchis) [reply process/track/sovereignBlockTrackerFactory.go:9]: moved to process/track/interface.go

File: process/track/interface.go [hunk 8]
```diff
@@ -47,3 +48,9 @@ type blockBalancerHandler interface {
 	SetLastShardProcessedMetaNonce(shardID uint32, nonce uint64)
 	IsInterfaceNil() bool
 }
+
+// BlockTrackerCreator is an interface for creating block trackers
+type BlockTrackerCreator interface {
+	CreateBlockTracker(argBaseTracker ArgShardTracker) (process.BlockTracker, error)
```
Comment (SebastianMarian) [inline process/track/interface.go:54]: argShardTracker

Comment (raduchis) [reply process/track/interface.go]: fixed

**Review by SebastianMarian — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**